### PR TITLE
Enable building Clash against GHC 9.10

### PR DESCRIPTION
**Pull request by christiaanb**
Add support for GHC 9.10

Currently, when building the Clash compiler against GHC 9.10 it will only work reliably on 64-bit platforms, and not so on 32-bit platforms. That's because we still use `Int` for our uniques, while GHC will use `Word64`, and we use `fromIntegral` to convert the `Word64` to `Int`. Users on 32-bit platforms should use GHC 9.8 or older to compile the Clash compiler for now.

Once this is merged, open an issue to rethink our `Unique` story so that it works with GHC's changeover from `Int` to `Word64` so that they have sufficient uniques on 32-bit platforms.

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

### CODE REVIEW

File: clash-cores/test/Test/Cores/Crc.hs [hunk 1]
```diff
@@ -61,7 +61,9 @@ proxyToSNat :: KnownNat n => Proxy n -> SNat n
 proxyToSNat _ = SNat
 
 nearestCleanDivisor :: Integer -> Integer -> Integer
-nearestCleanDivisor a b =  List.head $ List.sortOn (\i -> abs (i - b)) candidates
+nearestCleanDivisor a b = case List.sortOn (\i -> abs (i - b)) candidates of
```
Comment (rowanG077) [inline clash-cores/test/Test/Cores/Crc.hs:64]: Why was this changed? The list should never be empty. Simply ro silence a newly introduced `List.head` usage warning?

Comment (christiaanb) [reply clash-cores/test/Test/Cores/Crc.hs:64]: Yeah, head and tail give warnings as of base-4.19

File: clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs [hunk 2]
```diff
@@ -161,6 +161,20 @@ ghcTypeToHWType iw = go
                     stripVoid t = t
                 _ -> ExceptT (MaybeT (go reprs m arg0))
 
+        -- XXX: this is a hack to get a KnownDomain from a KnownConfiguration
+        "GHC.Classes.CTuple2"
+          | [arg0@(tyView -> TyConApp kdNm _), arg1] <- args
+          , nameOcc kdNm == "Clash.Signal.Internal.KnownDomain"
+          -> case tyView arg1 of
+                TyConApp kdNm1 _
+                  | nameOcc kdNm1 == "Clash.Signal.Internal.KnownDomain"
```
Comment (martijnbastiaan) [inline clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs:170]:
```
| nameOcc kdNm1 == show '' KnownDomain
```

File: clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs [hunk 3]
```diff
@@ -161,6 +161,20 @@ ghcTypeToHWType iw = go
...
+                  -> do k1 <- (stripVoid . stripFiltered) <$> ExceptT (MaybeT (go reprs m arg0))
+                        k2 <- (stripVoid . stripFiltered) <$> ExceptT (MaybeT (go reprs m arg1))
+                        returnN (Void (Just (Product "CTuple2" Nothing [k1,k2])))
```
Comment (martijnbastiaan) [inline clash-ghc/src-ghc/Clash/GHC/NetlistTypes.hs:173]: Shouldn't this be:

```
returnN (Void (Just (Product "GHC.Classes.CTuple2" Nothing [k1,k2])))
```

File: clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs [hunk 4]
```diff
@@ -4637,6 +4647,22 @@ ghcPrimStep tcm isSubj pInfo tys args mach = case primName pInfo of
     where
       op :: KnownNat n => Proxy n -> Double -> (Signed n, Double)
       op _ = properFraction
+  "GHC.Internal.Float.$wproperFractionDouble"
+    | _ : Lit (DoubleLiteral d) : _ <- args
+    , [sty@(tyView -> TyConApp signedTcNm [nTy@(LitTy (NumTy kn))])] <- tys
+    , nameOcc signedTcNm == "Clash.Sized.Internal.Signed.Signed"
```
Comment (martijnbastiaan) [inline clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:4672]:
```
, nameOcc signedTcNm == ''Signed
```

File: changelog/2024-07-26T07_36_58+02_00_ghc_910_upgrade [hunk 5]
```diff
@@ -0,0 +1,4 @@
+ADDED: Support for GHC 9.10.
+A word of caution: When the Clash compiler is compiled against GHC 9.10, it will
+currently only work reliably on 64-bit platforms. Compile the Clash compiler with
+GHC 9.8 or older if you are on a 32-bit platform.
```
Comment (martijnbastiaan) [inline changelog/2024-07-26T07_36_58+02_00_ghc_910_upgrade:4]: I guess we could add that to `BrokenGhcs` (even though its more like `BrokenClash` in this case :-)).

Comment (martijnbastiaan) [reply changelog/2024-07-26T07_36_58+02_00_ghc_910_upgrade:4]: Could you make an issue out of this and link it here?

**Review by martijnbastiaan — COMMENTED** [review]
I think we should add an issue to de-magic-ify names such as `GHC.Prim.dataToTagLarge#`. Right now it's really unclear what names actually exist in what GHC version.

In any case some small comments attached. Overall LGTM.

File: clash-lib/src/Clash/Netlist/BlackBox.hs [hunk 6]
```diff
@@ -564,6 +564,26 @@ mkPrimitive bbEParen bbEasD declType dst pInfo args tickDecls =
...
+            ["GHC.Prim.dataToTagSmall#", "GHC.Prim.dataToTagLarge#"] -> case args of
+              [Right _, Right _,Left (Data dc)] -> do
+                iw <- Lens.view intWidth
+                return (N.Literal (Just (Signed iw,iw)) (NumLit $ toInteger $ dcTag dc - 1),[])
+              [Right _, Right _,Left scrut] -> do
+                tcm      <- Lens.view tcCache
+                let scrutTy = inferCoreTypeOf tcm scrut
+                scrutHTy <- unsafeCoreTypeToHWTypeM' $(curLoc) scrutTy
+                (scrutExpr,scrutDecls) <-
+                  mkExpr False Concurrent (NetlistId (Id.unsafeMake "c$dtt_rhs") scrutTy) scrut
```
[the remark this answers is not in the document]
Comment (christiaanb) [reply clash-lib/src/Clash/Netlist/BlackBox.hs:577]: I copied what was there from from `dataToTag#` and didn’t give it a second thought really.

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Netlist/BlackBox.hs:577]: Ah ok, fair enough. Could you add an issue saying we should investigate this? In case someone gets an error with this ID it should point them to something tangible.

[the remark this answers is not in the document]
Comment (christiaanb) [reply clash-lib/src/Clash/Netlist/BlackBox.hs:577]: Tracked by https://github.com/clash-lang/clash-compiler/issues/2775

File: changelog/2024-07-26T07_36_58+02_00_ghc_910_upgrade (hunk 5 above)
Comment (christiaanb) [reply changelog/2024-07-26T07_36_58+02_00_ghc_910_upgrade:4]: Tracked by: https://github.com/clash-lang/clash-compiler/issues/2776

**Review by martijnbastiaan — APPROVED** [review]
Great!

File: clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs [hunk 7]
```diff
@@ -5884,5 +5937,5 @@ svoid :: (State# RealWorld -> State# RealWorld) -> IO ()
 svoid m0 = IO (\s -> case m0 s of s' -> (# s', () #))
 
 isTrueDC,isFalseDC :: DataCon -> Bool
-isTrueDC dc  = dcUniq dc == getKey trueDataConKey
-isFalseDC dc = dcUniq dc == getKey falseDataConKey
+isTrueDC dc  = dcUniq dc == fromIntegral (getKey trueDataConKey)
+isFalseDC dc = dcUniq dc == fromIntegral (getKey falseDataConKey)
```
Comment (leonschoorl) [inline clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:5941]: Instead of sprinkling around a very generic function like `fromIntegral`, I feel it might smart to define a conversion function from GHC uniques to clash uniques and use that everywhere.

Comment (leonschoorl) [reply clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:5941]: That function could for now just be `fromIntegral`, but if we need to change it in the future, we have a central place to do so.

File: clash-lib/src/Clash/Normalize/Transformations/Case.hs [hunk 8]
```diff
@@ -283,7 +284,9 @@ caseCon' ctx@(TransformContext is0 _) e@(Case subj ty alts) = do
       (Prim pInfo,repTy:_:msgOrCallStack:_,ticks)
         | primName pInfo `elem` ["Control.Exception.Base.patError"
                                 ,"GHC.Err.undefined"
-                                ,"GHC.Err.errorWithoutStackTrace"] ->
+                                ,"GHC.Err.errorWithoutStackTrace"
+                                ,"GHC.Internal.Err.undefined"
+                                ,"GHC.Internal.Err.errorWithoutStackTrace"] ->
```
Comment (leonschoorl) [inline clash-lib/src/Clash/Normalize/Transformations/Case.hs:289]: `parError` also moved to [GHC.Internal.Control.Exception.Base](https://hackage.haskell.org/package/ghc-internal-9.1001.0/docs/src/GHC.Internal.Control.Exception.Base.html#patError).

Unless the new ghc doesn't see that as a primitive anymore.

Comment (leonschoorl) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs:289]: With all these things moving maybe now is a good time to start using TemplateHaskellQuotes to get strings like these as broadly as possible.

Comment (christiaanb) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs:289]: I'd rather make a new issue to do that instead of doing it in this GHC 9.10 upgrade PR.

Comment (christiaanb) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs:289]: With regards to `patError`, I guess we do not have any tests in out testsuite that trigger it for GHC 9.10. Since it is a bottoming function, however, it will not have an unfolding, and will thus end up as a primitive. I'll check the other lists of "exceptions values" that we have to see whether I need to make some updates there as well.

File: clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs (hunk 7 above)
Comment (christiaanb) [reply clash-ghc/src-ghc/Clash/GHC/Evaluator/Primitive.hs:5941]: I'll make a function called `fromGhcUnique`.

File: clash-lib/src/Clash/Normalize/Transformations/Case.hs (hunk 8 above)
Comment (christiaanb) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs]: I've followed your advice for the bottoming values in https://github.com/clash-lang/clash-compiler/pull/2758/commits/73a811d1157f5eb06c190c12a1857a829e0c56a9

Comment (leonschoorl) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs]: Yes, I agree changing everything to use TH quotes it a big change and should be a different PR.
I've been thinking for a while that doing that would be a good change as it would make the code less fragile.

It just that while looking through this PR now seen like a good time because lots of stuff has moved.
But now that you already made the PR with the new names, the best time would have been "yesterday".

Comment (leonschoorl) [reply clash-lib/src/Clash/Normalize/Transformations/Case.hs]: With respect to `patError`, we have [the PatError test](https://github.com/clash-lang/clash-compiler/blob/1d49cd2a528fe719beb3b2887dc440362b1b3574/tests/shouldwork/Basic/PatError.hs) that should test this.

And (using 9.10) the GHC core has a `patError`:
```
topEntity
  = \ (ds_d1di :: Maybe Int) ->
      case ds_d1di of {
        __DEFAULT ->
          case GHC.Internal.Control.Exception.Base.patError
                 @GHC.Types.LiftedRep   
                 @()
                 "tests/shouldwork/Basic/PatError.hs:6:1-22|function topEntity"#
          of {};
        Just x_a1cs -> x_a1cs
      }
```

But not in a form that clash expects, using ghc 9.0 it looked like:
```
topEntity
  = \ (ds_dLU :: Maybe Int) ->
      case ds_dLU of {
        __DEFAULT ->
          Control.Exception.Base.patError
            @'GHC.Types.LiftedRep
            @Int
            "tests/shouldwork/Basic/PatError.hs:6:1-22|function topEntity"#;
        Just x_aGO -> x_aGO
      }
```

This new form with the empty case get converted by [GHC2Core into `Clash.Normalize.Primitives.undefined`](https://github.com/clash-lang/clash-compiler/blob/1d49cd2a528fe719beb3b2887dc440362b1b3574/clash-ghc/src-ghc/Clash/GHC/GHC2Core.hs#L516-L517):

```
Clash: Compiling PatError.topEntity
PatError.topEntity before normalization:

λ(ds :: GHC.Internal.Maybe.Maybe GHC.Types.Int) ->
case ds[LocalId] of
  GHC.Internal.Maybe.Nothing  ->
    Clash.Normalize.Primitives.undefined @GHC.Types.Int
  GHC.Internal.Maybe.Just (x :: GHC.Types.Int) ->
    x[LocalId]
```

File: clash-lib/src/Clash/Normalize/Primitives.hs-boot [hunk 9]
```diff
@@ -0,0 +1,6 @@
+module Clash.Normalize.Primitives where
+
+import Clash.Core.Term (PrimInfo)
+
+undefined :: PrimInfo
+undefinedX :: PrimInfo
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Normalize/Primitives.hs-boot:6]: Is it really worth it to add a `.hs-boot` file just to get the name of these two functions? I'd be inclined to make an exception for these and just refer to them as a string, in the interest of compilation speed.

**Review by DigitalBrains1 — COMMENTED** [review]
Overall, LGTM, but maybe we should improve the test for `PatError`? It looks like both GHC 9.10 and GHC 9.6 generate correct code for the test in the test suite, but it still seems like the test is insufficient even though it happens to have no repercussions here.

I did notice a typo in a commit msg:

> This also introduces a new functions, fromGhcUnique, which [...]

_functions_ should be _function_, singular. The question is, do we want to burn more CI energy on that? It _is_ something that will be there forever.

File: clash-lib/src/Clash/Normalize/Primitives.hs-boot (hunk 9 above)
Comment (christiaanb) [reply clash-lib/src/Clash/Normalize/Primitives.hs-boot:6]: I'd rather keep everything uniform as TH names.

**Comment by christiaanb** [comment]
> Overall, LGTM, but maybe we should improve the test for `PatError`? It looks like both GHC 9.10 and GHC 9.6 generate correct code for the test in the test suite, but it still seems like the test is insufficient even though it happens to have no repercussions here.
> 
> I did notice a typo in a commit msg:
> 
> > This also introduces a new functions, fromGhcUnique, which [...]
> 
> _functions_ should be _function_, singular. The question is, do we want to burn more CI energy on that? It _is_ something that will be there forever.

Given that the Core created in GHC 9.8 is also
```
[GblId,
 Arity=1,
 Str=<1L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [30] 200 0}]
topEntity
  = \ (ds_d1cV :: Maybe Int) ->
      case ds_d1cV of {
        Nothing ->
          case Control.Exception.Base.patError
                 @GHC.Types.LiftedRep
                 @()
                 "tests/shouldwork/Basic/PatError.hs:6:1-22|function topEntity"#
          of wild1_00 {
          };
        Just x_a1c5 -> x_a1c5
      }
```
which Clash's GHC2Core turns into:
```
λ(ds[7205759403792798227] :: GHC.Maybe.Maybe[3674937295934324792]
                               GHC.Types.Int[3674937295934324766]) ->
case ds[7205759403792798227][LocalId] of
  GHC.Maybe.Nothing[3891110078048108565]  ->
    Clash.Normalize.Primitives.undefined
      @GHC.Types.Int[3674937295934324766]
  GHC.Maybe.Just[3891110078048108568]
    (x[6989586621679014385] :: GHC.Types.Int[3674937295934324766]) ->
    x[6989586621679014385][LocalId]
```
The `PatError` test being broken is not a result of this upgrade to GHC 9.10; but was already present. I'll there make a separate issue linking to the discussion here.
